presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/alluxio/AlluxioProtoUtils.java

```diff
@@ -0,0 +1,284 @@
+                .setComment(database.hasComment() ? Optional.of(database.getComment()) : Optional.empty())
+                .setParameters(database.getParameterMap())
+                .build();
+    }
+
+    public static Table fromProto(alluxio.grpc.table.TableInfo table)
+    {
+        if (!table.hasLayout()) {
+            throw new UnsupportedOperationException("Unsupported table metadata. missing layout.");
+        }
+        Layout layout = table.getLayout();
+        if (!alluxio.table.ProtoUtils.isHiveLayout(layout)) {
+            throw new UnsupportedOperationException("Unsupported table layout: " + layout);
+        }
+
+        try {
+            PartitionInfo partitionInfo = alluxio.table.ProtoUtils.toHiveLayout(layout);
+
+            // compute the data columns
+            Set<String> partitionColumns = table.getPartitionColsList().stream().map(FieldSchema::getName).collect(toImmutableSet());
+            List<FieldSchema> dataColumns = table.getSchema().getColsList().stream().filter((field) -> !partitionColumns.contains(field.getName())).collect(toImmutableList());
+
+            Table.Builder builder = Table.builder()
+                    .setDatabaseName(table.getDbName())
+                    .setTableName(table.getTableName())
+                    .setOwner(table.getOwner())
+                    // TODO - Alluxio should return the actual table type
+                    .setTableType(PrestoTableType.OTHER)
+                    .setDataColumns(dataColumns.stream().map(AlluxioProtoUtils::fromProto).collect(toImmutableList()))
+                    .setPartitionColumns(table.getPartitionColsList().stream().map(AlluxioProtoUtils::fromProto).collect(toImmutableList()))
+                    .setParameters(table.getParametersMap())
+                    .setViewOriginalText(Optional.empty())
+                    .setViewExpandedText(Optional.empty());
+            alluxio.grpc.table.layout.hive.Storage storage = partitionInfo.getStorage();
+            // TODO: We should also set storage parameters here when they are available in alluxio.grpc.table.layout.hive.Storage
+            builder.getStorageBuilder()
+                    .setSkewed(storage.getSkewed())
+                    .setStorageFormat(fromProto(storage.getStorageFormat()))
+                    .setLocation(storage.getLocation())
+                    .setBucketProperty(storage.hasBucketProperty() ? fromProto(storage.getBucketProperty()) : Optional.empty())
+                    .setSerdeParameters(storage.getStorageFormat().getSerdelibParametersMap());
+            return builder.build();
+        }
+        catch (InvalidProtocolBufferException e) {
+            throw new IllegalArgumentException("Failed to extract PartitionInfo from TableInfo", e);
+        }
+    }
+
+    public static HiveColumnStatistics fromProto(ColumnStatisticsData columnStatistics, OptionalLong rowCount)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 34. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks such as on lines 132-139, lines 140-147, lines 148-155, lines 156-163, lines 164-176 into separate methods.